[CI, enhancement] add pytest-cov for testing

icfaust opened the pull request:
## Description

Add the possibility to run pytest-cov for test coverage reports for CI, and add them to the Github Actions CI for PRs.  There will be a follow-up PR if we chose to integrate this into coveralls.io or codecov based on internal discussions.

How to use these files: 

- navigate to your GitHub Actions CI run (for example: https://github.com/intel/scikit-learn-intelex/actions/runs/11851114955)
- Download and extract all files from all artifacts
- Install coverage.py :https://coverage.readthedocs.io/en/7.6.5/
- run: ```coverage combine /coverage_lnx_Py3.10_1.2 /coverage_lnx_Py3.9_1.1 /coverage_lnx_Py3.11_1.3 /coverage_win_Py3.10_1.2 /coverage_win_Py3.9_1.1 /coverage_win_Py3.11_1.3``` https://coverage.readthedocs.io/en/latest/cmd.html#combining-data-files-coverage-combine
- Generate report of your choice example: ```coverage report``` https://coverage.readthedocs.io/en/latest/cmd.html#reporting
---

Checklist to comply with **before moving PR from draft**:

**PR completeness and readability**

- [x] I have reviewed my changes thoroughly before submitting this pull request.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] I have updated the documentation to reflect the changes or created a separate PR with update and provided its number in the description, if necessary.
- [x] Git commit message contains an appropriate signed-off-by string _(see [CONTRIBUTING.md](https://github.com/intel/scikit-learn-intelex/blob/main/CONTRIBUTING.md#pull-requests) for details)_.
- [x] I have added a respective label(s) to PR if I have a permission for that.
- [x] I have resolved any merge conflicts that might occur with the base branch.

**Testing**

- [x] I have run it locally and tested the changes extensively.
- [x] All CI jobs are green or I have provided justification why they aren't.
- [x] I have extended testing suite if new functionality was introduced in this PR.

**Performance**

- [x] I have measured performance for affected algorithms using [scikit-learn_bench](https://github.com/IntelPython/scikit-learn_bench) and provided at least summary table with measured data, if performance change is expected.
- [x] I have provided justification why performance has changed or why changes are not expected.
- [x] I have provided justification why quality metrics have changed or why changes are not expected.
- [x] I have extended benchmarking suite and provided corresponding scikit-learn_bench PR if new measurable functionality was introduced in this PR.

samir-nasibli reviewed (commented):
Thank you @icfaust for that important tool integration.
I will take a look today later.

ahuber21 commented on .ci/scripts/run_sklearn_tests.py:
```diff
@@ -37,22 +37,38 @@
 
-    os.chdir(os.path.dirname(sklearn.__file__))
+    cwd = os.getcwd()
+    root_dir = os.path.dirname(sklearn.__file__)
```
```suggestion
    sklearn_file_dir = os.path.dirname(sklearn.__file__)

```

ahuber21 commented on .ci/scripts/run_sklearn_tests.py:
```diff
@@ -37,22 +37,38 @@
         f'{os.environ["DESELECTED_TESTS"]} {os.environ["SELECTED_TESTS"]}'.split(" ")
     )
+
+    if os.getenv("COVERAGE_RCFILE"):
+        pytest_args += (
+            f'--cov=onedal --cov=sklearnex --cov-config={os.environ["COVERAGE_RCFILE"]} '
+            "--cov-report=term".split(" ")
+        )
```
```suggestion
    if rc := os.getenv("COVERAGE_RCFILE"):
        pytest_args += (
            f'--cov=onedal --cov=sklearnex --cov-config={rc} '
            "--cov-report=term".split(" ")
        )

```

Since we no longer support python < 3.8 we can start using the walrus operator.

ahuber21 commented on .ci/scripts/run_sklearn_tests.py:
```diff
@@ -37,22 +37,38 @@
 
     if args.device != "none":
         with sklearn.config_context(target_offload=args.device):
-            return_code = pytest.main(pytest_args)
```
Why was the conversion added? What were non-int returns previously?

icfaust replied:
Conversion to int occurs at the end of the file, I wanted to have it earlier to check for a successful run, so I moved it into the return code.

ahuber21 commented on onedal/basic_statistics/tests/test_basic_statistics.py:
Please provide an explanation for these changes

icfaust replied:
Good question, it was the only test file getting picked up for code coverage because of this inclusion, which other testing files are dependent on.  I simplified it and moved it into a location which was more conducive for wider use (onedal.tests.utils).  Let me know what you think.

david-cortes-intel commented:
@icfaust Would be nice to get an HTML report in addition to plain text.

david-cortes-intel commented on .github/workflows/ci.yml:
```diff
@@ -188,15 +200,13 @@ jobs:
-          echo "DPCFLAG=${DPCFLAG}" >> "$GITHUB_OUTPUT"
+          echo C:\msys64\usr\bin;>> %GITHUB_PATH%
+          echo NO_DIST=YES>> %GITHUB_ENV%
+          set DPCTL_TEMP="${{ env.DPCTL_PY_VERSIONS }}"
```
If this is for windows, shouldn't the quote start before the name of the variable that you are setting?

icfaust replied:
Good catch, luckily in this case the quotation marks do not impact the operation of the underlying check. Will swap it nevertheless

icfaust replied:
Turns out changing this interprets the ```|``` into an operation and will fail. Reverted to original.

david-cortes-intel commented on requirements-test.txt:
```diff
@@ -1,6 +1,7 @@
 pytest==7.4.4 ; python_version <= '3.10'
 pytest==8.3.3 ; python_version >= '3.11'
 pytest-json-report==1.5.0
+pytest-cov==5.0.0
```
Last version is 6.0 - why use an older one?

icfaust replied:
Good catch, the version 6.0 was released as this PR was waiting for review. Will update.

david-cortes-intel commented on .ci/scripts/run_sklearn_tests.py:
```diff
@@ -37,16 +37,25 @@
+
+    if rc := os.getenv("COVERAGE_RCFILE"):
+        pytest_args += (
+            f"--cov=onedal --cov=sklearnex --cov-config={rc} "
+            "--cov-report=term".split(" ")
```
```suggestion
            "--cov=onedal", "--cov=sklearnex", f"--cov-config={rc}",
            "--cov-report=term",

```

icfaust replied:
removed term as well since its assumed that the reports will be viewed only after integration together.

david-cortes-intel commented on conda-recipe/run_test.sh:
```diff
@@ -60,16 +70,16 @@ return_code=$(($return_code + $?))
 return_code=$(($return_code + $?))
 
-pytest --verbose -s "${sklex_root}/.ci/scripts/test_global_patch.py" $@ $(json_report_name global_patching)
+pytest --verbose -s ${sklex_root}/.ci/scripts/test_global_patch.py $@ $(json_report_name global_patching) "${COV_ARGS[@]}"
```
I think without the quotes it'd end up failing if there spaces in the path.

icfaust replied:
Oh shoot, that was definitely an oversight. I'll make sure the quotes are readded.

david-cortes-intel commented on .github/workflows/ci.yml:
```diff
@@ -197,15 +209,13 @@ jobs:
-          if [ $(echo ${{ matrix.PYTHON_VERSION }} | grep '${{ env.DPCTL_PY_VERSIONS }}') ]; then DPCFLAG=''; fi
-          echo "DPCFLAG=${DPCFLAG}" >> "$GITHUB_OUTPUT"
+          echo C:\msys64\usr\bin;>> %GITHUB_PATH%
+          echo NO_DIST=YES>> %GITHUB_ENV%
```
We're using `NO_DIST=1` elsewhere.

icfaust replied:
changed.

icfaust replied:
Turns out it broke github actions. Had to revert. Why? Because exporting to GITHUB_ENV via a batch file seems very poorly documented/ operates in an unexpected manner.  Would do it in linux, but I need the full path of the coverage rc file from windows, and running a bash script from github actions forces it into a linux environment (another odd behavior).

david-cortes-intel replied:
But why would a '1' in this case make such a difference? Or was it about a different change?

icfaust replied:
Something with how github is writing and reading to the GITHUB_ENV file is dropping the 1 when done from a batch script. Quotations or making it a string with "" will keep the quotations too. Valid strings for the input can be found here: https://github.com/intel/scikit-learn-intelex/blob/main/setup.py#L85 so it should be okay.

samir-nasibli commented:
> ## Description
> Add the possibility to run pytest-cov for test coverage reports for CI, and add them to the Github Actions CI for PRs.
> 
> Checklist to comply with **before moving PR from draft**:
> 
> **PR completeness and readability**
> 
> * [x]  I have reviewed my changes thoroughly before submitting this pull request.
> * [x]  I have commented my code, particularly in hard-to-understand areas.
> * [x]  I have updated the documentation to reflect the changes or created a separate PR with update and provided its number in the description, if necessary.
> * [x]  Git commit message contains an appropriate signed-off-by string _(see [CONTRIBUTING.md](https://github.com/intel/scikit-learn-intelex/blob/main/CONTRIBUTING.md#pull-requests) for details)_.
> * [x]  I have added a respective label(s) to PR if I have a permission for that.
> * [x]  I have resolved any merge conflicts that might occur with the base branch.
> 
> **Testing**
> 
> * [x]  I have run it locally and tested the changes extensively.
> * [x]  All CI jobs are green or I have provided justification why they aren't.
> * [x]  I have extended testing suite if new functionality was introduced in this PR.
> 
> **Performance**
> 
> * [x]  I have measured performance for affected algorithms using [scikit-learn_bench](https://github.com/IntelPython/scikit-learn_bench) and provided at least summary table with measured data, if performance change is expected.
> * [x]  I have provided justification why performance has changed or why changes are not expected.
> * [x]  I have provided justification why quality metrics have changed or why changes are not expected.
> * [x]  I have extended benchmarking suite and provided corresponding scikit-learn_bench PR if new measurable functionality was introduced in this PR.

Could you please add to the description how to use this?
If this already integrated into CI, could you please share link for statistics?

icfaust commented:
/intelci: run

icfaust commented:
/intelci: run

icfaust commented:
> > ## Description
> > Add the possibility to run pytest-cov for test coverage reports for CI, and add them to the Github Actions CI for PRs.
> > Checklist to comply with **before moving PR from draft**:
> > **PR completeness and readability**
> > 
> > * [x]   I have reviewed my changes thoroughly before submitting this pull request.
> > * [x]   I have commented my code, particularly in hard-to-understand areas.
> > * [x]   I have updated the documentation to reflect the changes or created a separate PR with update and provided its number in the description, if necessary.
> > * [x]   Git commit message contains an appropriate signed-off-by string _(see [CONTRIBUTING.md](https://github.com/intel/scikit-learn-intelex/blob/main/CONTRIBUTING.md#pull-requests) for details)_.
> > * [x]   I have added a respective label(s) to PR if I have a permission for that.
> > * [x]   I have resolved any merge conflicts that might occur with the base branch.
> > 
> > **Testing**
> > 
> > * [x]   I have run it locally and tested the changes extensively.
> > * [x]   All CI jobs are green or I have provided justification why they aren't.
> > * [x]   I have extended testing suite if new functionality was introduced in this PR.
> > 
> > **Performance**
> > 
> > * [x]   I have measured performance for affected algorithms using [scikit-learn_bench](https://github.com/IntelPython/scikit-learn_bench) and provided at least summary table with measured data, if performance change is expected.
> > * [x]   I have provided justification why performance has changed or why changes are not expected.
> > * [x]   I have provided justification why quality metrics have changed or why changes are not expected.
> > * [x]   I have extended benchmarking suite and provided corresponding scikit-learn_bench PR if new measurable functionality was introduced in this PR.
> 
> Could you please add to the description how to use this? If this already integrated into CI, could you please share link for statistics?

I'll add to the description, but as of now all it will do is store the necessary artifacts for use locally on your dev machine or via a service like https://about.codecov.io/ or https://coveralls.io/  That will come after necessary internal discussion of which direction to go.

Vika-F commented on sklearnex/linear_model/logistic_regression.py:
```diff
@@ -289,6 +289,7 @@ def _onedal_fit(self, X, y, sample_weight=None, queue=None):
             if queue is None or queue.sycl_device.is_cpu:
                 return self._onedal_cpu_fit(X, y, sample_weight)
 
+            #  coverage: gpu_start
```
What is the purpose of those comments? And why are they added only for LogReg?

icfaust replied:
Good question! We can avoid code coverage analysis for regions between comments ```coverage: gpu_start``` and ```coverage: gpu_stop``` since we do not do any GPU operations in public CI. If someone does a full GPU-only PR they will see a large decrease in code coverage, which we can designate as GPU code.  Do to so, we need to remove the ```#``` from this file: https://github.com/icfaust/scikit-learn-intelex/blob/dev/pytest-cov/.coveragerc#L23

icfaust replied:
Oh, and I did it only for LogisticRegression as an initial thing because the sklearnex code is only GPU-relevant, as CPU is offloaded to daal4py (and would otherwise show very bad code-coverage metrics).

Vika-F replied:
Thanks for the explanation.
But it looks like a lot of work to write all those comments properly in the code and maintain them in good quality.
Can there be a better way to exclude GPU-related code from the code coverage analysis?

Vika-F replied:
Maybe it is possible to add something like:
```
exclude_also =
    self\._onedal_estimator\..*queue=queue\)
```
to .coveragerc. To exclude all the estimators calls that use SYCL queue.

icfaust replied:
Since we aren't making a hard and fast rule, it could be that we make the first round of analysis without gpu exclusion, the metrics will show regressions with GPU code introduction, but the way GPU code isn't fully standardized in the codebase, so you are right that it would be a long a tedious process.

icfaust replied:
I'll start by removing the secondary excludes from LogisticRegression and elsewhere, and a development ticket to the backlog (let me know what you think @Vika-F ).  I think this will occur after we integrate into coveralls.

Vika-F commented on onedal/tests/utils/__init__.py:
```diff
@@ -0,0 +1,51 @@
+    pass_if_not_implemented_for_gpu,
+)
+
+basicstats_options_and_tests = {
```
Is it possible to place this definition into `basic_statistics/tests` module?
Because `utils` does not seem to be a proper place for this kind of definitions.

I like the overall basic statistics tests refactoring, it's just the placement of this dictionary looks not so good to me.

icfaust replied:
I'll move it an add another omission to the coverage.rc file for that case, otherwise it will be picked up in coverage metrics.

icfaust commented on .coveragerc:
```diff
@@ -0,0 +1,24 @@
+[run]
```
Done! Will address the other issue tomorrow.

ethanglaser reviewed (commented):
Its looking good - can you remind what the conclusion of how this info would be used? Like would we have a check that sees if coverage worsens with changes or just manual analysis for now?

ethanglaser commented:
> Is there an example of manually generated report for this?

From "Archive coverage data" step in public CI there is https://github.com/intel/scikit-learn-intelex/actions/runs/11838651610/artifacts/2187948371 for example, not sure the best way to view the contents though

icfaust commented on sklearnex/basic_statistics/tests/test_incremental_basic_statistics.py:
The coverage metrics were picking up this test file for coverage because of how functionality was being re-used, it was moved to a separate file so that test coverage would not be generated for test functions.

icfaust commented on .github/workflows/ci.yml:
```diff
@@ -261,15 +271,25 @@ jobs:
+          name: coverage_win_Py${{ matrix.PYTHON_VERSION }}_${{ matrix.SKLEARN_VERSION }}
+          path: |
+             .coverage.sklearn
+             .coverage.sklearnex
```
There will be two coverage output files following coverage.py's convention (see this example: https://coverage.readthedocs.io/en/latest/cmd.html#combining-data-files-coverage-combine). This is not done yet in this PR, but all the various coverage files will be combined and then can be used to generate reports: https://coverage.readthedocs.io/en/latest/cmd.html#reporting.  Ideally we would integrate into https://coveralls.io/ (what IntelPython uses) or https://about.codecov.io/ (what sklearn uses) which will automatically calculate coverage values and will add comments in PRs.  We could create HTML files for seeing which lines of a PR are unused, but public CI doesn't contain any GPU code testing, and as of now GitHub Actions CI doesn't test our supported sklearn versions.  So long story short: which way would you @ethanglaser  like to take it? Ideally I would want to do it in a follow-up PR.

ethanglaser replied:
All sounds good - it would be interesting to get something set up with python team to see how they are using this

icfaust commented on .ci/scripts/run_sklearn_tests.py:
```diff
@@ -37,16 +37,27 @@
     args = parser.parse_args()
 
-    os.chdir(os.path.dirname(sklearn.__file__))
+    cwd = os.getcwd()
```
removed

icfaust commented on onedal/tests/utils/__init__.py:
removed from the code, reverted to previous.

icfaust commented:
/intelci: run

icfaust commented:
/intelci: run

icfaust commented:
/intelci: run

icfaust commented on conda-recipe/run_test.bat:
```diff
@@ -28,6 +28,8 @@ if "%PYTHON%"=="python" (
     set NO_DIST=1
 )
 
+IF DEFINED COVERAGE_RCFILE (set COV_ARGS=--cov=onedal --cov=sklearnex --cov-config=%COVERAGE_RCFILE% --cov-append --cov-report=) else (set COV_ARGS=)
```
done, also changed ``` python``` above it

david-cortes-intel replied:
I think the `defined` call requires having the variable name without the percent signs, otherwise it will apply substitutions on it and the statement might not work. Also looks like the double quotes were lost.

icfaust replied:
Changes as requested made.

david-cortes-intel commented on conda-recipe/run_test.bat:
```diff
@@ -20,14 +20,16 @@ rem %1 - scikit-learn-intelex repo root (should end with '\', leave empty if it'
 
 setlocal enableextensions
-IF NOT DEFINED PYTHON (
+IF NOT DEFINED %PYTHON% (
```
```suggestion
IF NOT DEFINED PYTHON (
```

david-cortes-intel commented on conda-recipe/run_test.bat:
```diff
@@ -20,14 +20,16 @@ rem %1 - scikit-learn-intelex repo root (should end with '\', leave empty if it'
     set NO_DIST=1
 )
 
+IF DEFINED %COVERAGE_RCFILE% (set COV_ARGS=--cov=onedal --cov=sklearnex --cov-config=%COVERAGE_RCFILE% --cov-append --cov-report=) else (set COV_ARGS=)
```
```suggestion
IF DEFINED COVERAGE_RCFILE (set COV_ARGS=--cov=onedal --cov=sklearnex --cov-config="%COVERAGE_RCFILE%" --cov-append --cov-report=) else (set COV_ARGS=)
```

icfaust commented:
/intelci: run

icfaust commented on .coveragerc:
```diff
@@ -0,0 +1,39 @@
+onedal =
+    onedal/
+    */onedal
+#[report]
+#exclude_also =
+#    def _onedal_gpu.*
+#    coverage: gpu_start(?s:.)*coverage: gpu_stop
```
@Alexsandruss done. please review

david-cortes-intel commented on conda-recipe/run_test.sh:
```diff
@@ -48,28 +48,34 @@ if [[ "$*" == *"--json-report"* ]]; then
+    if [ -n "${COVERAGE_RCFILE}" ]; then
+        ARGS+=(--cov=onedal --cov=sklearnex --cov-config="${COVERAGE_RCFILE}" --cov-append --cov-report=)
     fi
+    "${ARGS[@]}"
```
I think this doesn't do the same thing. It would end up trying to execute a command instead of adding arguments to an existing call.

icfaust replied:
good catch, we should schedule a meeting or make a chat to iron out these issues quickly. i have no particular preference

david-cortes-intel commented on conda-recipe/run_test.sh:
```diff
@@ -48,28 +48,34 @@ if [[ "$*" == *"--json-report"* ]]; then
+    if [ -n "${COVERAGE_RCFILE}" ]; then
+        ARGS+=(--cov=onedal --cov=sklearnex --cov-config="${COVERAGE_RCFILE}" --cov-append --cov-report=)
     fi
+    printf -- "${ARGS[@]}"
```
I think this ends up adding only the first argument.

david-cortes-intel commented on conda-recipe/run_test.sh:
```diff
@@ -48,28 +48,34 @@ if [[ "$*" == *"--json-report"* ]]; then
+    if [ -n "${COVERAGE_RCFILE}" ]; then
+        ARGS+=(--cov=onedal --cov=sklearnex --cov-config="${COVERAGE_RCFILE}" --cov-append --cov-report=)
     fi
+    printf -- "${ARGS[@]}"
```
```suggestion
    printf -- "${ARGS[*]}"
```

icfaust replied:
Looks like the change isn't needed. It produced the requisite coverage files

david-cortes-intel replied:
Does it still produce them if you add both the json report argument and the coverage argument?

icfaust replied:
@david-cortes-intel I checked the file and you are right that it is incomplete, I have changed it to ```"${ARGS[*]}"``` along with some other necessary changes. I will not merge without your approval because of the changes to the report generation lines.

icfaust replied:
@david-cortes-intel I have added a matching test to the run_test.sh to match run_test.bat for checking file generation. I hope that will help address the file generation in linux question.

ethanglaser commented on onedal/basic_statistics/tests/utils.py:
This is much cleaner than what we have in main branch. Nice.

ethanglaser reviewed (approved):
No objections from my side assuming private CI is good and comments around bash/bat/yml syntax will be resolved

ethanglaser commented:
/intelci: run

david-cortes-intel commented on conda-recipe/run_test.bat:
```diff
@@ -28,54 +28,45 @@ if "%PYTHON%"=="python" (
 
+set "PYTEST_ARGS= "
+
+IF DEFINED COVERAGE_RCFILE (set PYTEST_ARGS=--cov=onedal --cov=sklearnex --cov-config="%COVERAGE_RCFILE%" --cov-append --cov-report= %PYTEST_ARGS%)
```
Actually on second thought, I think these double quotes around `%COVERAGE_RCFILE%` would have an unintended effect of leaving quotes in the argument, so perhaps it should either be done without the quotes, or with the quotes around the whole `set "var=value"` call.

Sorry for the confusion.

icfaust replied:
No worries, have made each of the PYTEST_ARGS in quotes as you've suggested (on next commit)

david-cortes-intel commented on conda-recipe/run_test.bat:
```diff
@@ -28,54 +28,45 @@ if "%PYTHON%"=="python" (
     del /q .pytest_reports\*.json
 )
 
 echo "NO_DIST=%NO_DIST%"
```
Maybe it could print the `PYTEST_ARGS` template here.

icfaust replied:
Added!

david-cortes-intel commented on conda-recipe/run_test.bat:
```diff
@@ -28,54 +28,45 @@ if "%PYTHON%"=="python" (
-        if !errorlevel! NEQ 0 (
-            set exitcode=1
-        )
+pytest --verbose -s "%1tests" %PYTEST_ARGS:FILENAME=legacy_report%|| set exitcode=1
```
I'm not sure here if having a space before `||` makes a difference or not, but I think it'd be better to leave it just in case, just for clarity that it's not part of the argument.

icfaust replied:
Sounds good, I will let CI continue running to verify things as they are and will make changes combined with whatever failures pop up.

david-cortes-intel reviewed (approved):
LGTM, just an ammendment about my earlier comment around the double quotes in the .bat file.
